# ENGR222 Lab 1 - A brief introduction to Python3, matplotlib and numpy

**Brendan Harding, Victoria University of Wellington**

**Last updated: 26th Feb 2020**


## Introduction

Welcome to your first ENGR222 computer lab.
The aim of this lab is to introduce you to Python3 and some of the modules you will use throughout this course.
Python is an extremely flexible and extensible programming language that can be used for almost any concievable purpose.
The focus of these lab sessions will be using Python for scientific computing, it is not intended to be a comprehensive guide to programming in Python.
There are three modules that will be used throughout the lab sessions, numpy, scipy and matplotlib.
Additionally, you are currently reading this within a jupyter notebook via the jupyter module.
Jupyter is a great tool that will allow us to program with python in an interactive and dynamic manner.

Note: this particular cell is a *markdown* cell, as opposed to a code cell, which allows me to write nicely formatted text. It is very useful for me to use when writing up these tutorial notes but it is not something that you, the student, need to be familiar with. 
You will use and interact with *code* cells, which are the default cell type within jupyter notebooks.

## Using Python as a calculator

At the most basic level, you can use Python as a calculator.
For example, in the following cell, try some basic calculations like `1+2`, `3.14-2.71`, `5*9`, `2.3*5.1`, `21.3/4.7` etc.
Observe that parentheses/brackets can be used as you normally would with a calculator e.g. `(6.4+9.7)/(4.3-1.7)`.
Simply type what you want to calculate, then hold **shift** and press **return** (or **enter**) to execute that cell.

In [ ]:
(4.7+8.2)*6.5

You can also raise numbers to integer powers, or even perform more general exponentiation, by using `**` (i.e. `**` is the Python version of `^` on a traditional calculator, in Python `^` is reserved for a different operation, the logical xor operation specifically).
Try calculating say `2**5`, `3.14**2`, `10**0.5`, etc.

In [ ]:
10**0.5

Notice if you have two lines of calculations in a cell only the result of the last line is displayed.
If the results of multiple calculations are to be displayed in one cell then we can use the standard `print` function.
Compare the following two cells.

In [ ]:
1+2
3+4

In [ ]:
print(1+2)
print(3+4)

The `print` function is quite flexible and can be used to display multiple results, and can print *strings* as well.
A *string* is essentially just text data that is enclosed within either single or double quotes.
Here is a couple of examples.
Play around and come up with some more.

In [ ]:
print(1+2,3+4)
print('The result of 1+2 is',1+2)

At this stage you probably want to ask: how do you obtain value of special numbers like $\pi$ or $e$, or want to use a special function like $\sin(x),\cos(x),\tan(x),\log(x),e^{x}$ etc?

These are all contained in the standard **math** module which is included in python by default.
To use these you first need to import the math module using `import math` and you can then calculate something like `math.sin(math.pi/3.0)`. Try this out in the cell below.

In [ ]:
import math
math.sin(math.pi/3.0)

Note that importing a module only needs to be performed once in a session. In the cell below note that you can continue to use math routines without importing again (provided you have executed the cell above containing `import math`, Python can only know about previous cells if they have been executed in the current session). Try some different calculations.

*Hint: if you want to know what functions are available in the math module they are documented [here](https://docs.python.org/3/library/math.html).*

In [ ]:
math.log(math.exp(3.0))

Typically, the very first cell of a notebook should import all of the common modules you will use throughout and is something you execute first thing when you open (or restart) a notebook.

Additionally, writing `math.` every time we want a function from the math module is a pain. 
There are several alternative ways to load/use modules.
One option is to import the module and assign it an alias, for example `import math as m` where here **m** becomes and alias to the math module so that we can call functions within the math module using `m.sin(3.0)` for example, thereby saving us from typing three character!
Specific functions can be imported using for example `from math import sin` which would allow us to just call `sin(3.0)` directly rather than `math.sin(3.0)`.
If we want lots of functions we could import them as `from math import sin,cos,tan` and so on, but this too can become tediuous.
If we want to import everything from the math module we can simply write `from math import *` and then any function in the math module can be called directly.

The last option is very convenient, but can also be problematic if you load several modules in this way which contain functions which have the same name.
I won't discuss this further but it is something to keep in mind.
The following cells illustrate this method of module loading.

In [ ]:
from math import *

In [ ]:
log(exp(pi))

Note: Most of the functions in the `math` module are also available in the `numpy` module.
Throughout this course the `numpy` module will be preferred, for reasons that will be discussed later.

## Commenting your code

Adding comments to code is a very useful way document, or add notes to, your code.
It is good practice to add comments to explain anything that may be unclear.
In Python, comments can be placed at the end of a line by adding a hashtag `#`.
Anthing that follows the hashtag on a single line is not executed.
Jupyter will also display it in gray to distinguish it from the code.
The following cell shows an example, albeit with some relatively long comments.

In [ ]:
1+2 # This result will not be printed as it is not the last line of the cell
print(3+4) # This result will be printed as it is wrapped in the print() function
5+6 # This result will be printed as it is the last line of the cell
# Comment on a line of it's own
# and extending onto a second line

## Assiging and using variables

Numbers can be assigned to variables (in fact any object you might deal with in Python can be assigned to a variable).
Variable names can be almost any string of letters and numbers, and may also contain underscores. 
There are exceptions where certain words are reserved for the Python programming language itself, some you will see later. 
Usefully, such words are typically displayed in another colour within jupyter so you can see they already have a meaning. 
Another exception is a variable cannot begin with a number.
Some examples of assigning numbers to variables and using them in calculations is provided in the next cell.
You should experiment further with this.

In [ ]:
a = 0.1
b = 2*pi
c = a+b # assign c to be the sum of a and b
ab = a*b # assign to ab the product of a and b
print(a,b,c,ab) # print all four variables defined above
a = 0.2 # replaces the definition of a
print(a) # see
long_variable_name_with_underscores = -4.7
AnotherLongVariableName = 2.1
x2 = log(5.0) # variable name with a number
x3y = 9.1 # another variable name with a number

Beware that definitions of imported functions or variables get overwritten if you use them as a variable name.
See the following examples.

In [ ]:
pi = 3.14 # Beware! Replaces pi variable obtained from 'from math import *'
print(pi) # see
cos = 0.5 # Beware! Replaces cos function obtained from 'from math import *'
cos(2.0) # Raises a TypeError because cos is now a number (0.5) and not a function

When a variable is just a number, you do some arithmetic on it *in-place* using `-=,+=,*=,/=`.
Play around with the following examples.

In [ ]:
a = 1.0
print(a)
a *= 2.5 # multiply a by 2.5
print(a)
a += 17.5 # add 17.5 to a
print(a)
a /= 5.0 # divide a by 5
print(a)
a -= 6.0 # subtract 6 from a
print(a)

## Numbers types in Python

Most computations performed in Python will use (double precision) floating point numbers.
These numbers are a *finite* subset of the *rational* numbers.

It is important to realise that floating point numbers have a limited accuracy. 
Moreover, as you continue to iterate computations on numbers with limited accuracy it is typical for the errors to accumulate and grow over time.
There are whole fields of study devoted to understanding this and designing numerical algorithms which prevents this error from blowing up so large that the results of computations are meaningless (this is referred to as numerical stability).

For now, it is enough to note that a (double precision) floating point number is only good up to about 15 to 16 decimal digits, and typically Python will print results with this may digits unless the result can be exactly represented with less digits.

Floating point numbers are internally represented using a binary format.
To represent a number in floating point format it is first decomposed (exactly) as 
$$
(\text{floating point number}) = (\text{sign})\times 2^{(\text{exponent})} \times (\text{mantissa})
$$
where the exponent and sign are (uniquely) chosen so that the mantissa lies in the interval $[1,2)$.
A double precision floating point number has 64 bits (binary digits) to store this representation.
One bit is used to denote the sign of the number (i.e. whether it is positive with sign$=+1$, or negative with sign$=-1$). 
Some more bits are used to represent the exponent (an integer expressed in base 2). 
Remaining bits *approximate* the mantissa (in a binary representation).

You don't really need to know the specific details beyond this for this course, but there are some consequences of this fact that may surprise you.
For example, $0.1$ may seem like a nice simple number that a computer should be able to represent exactly.
However, an exact floating point representation of $0.1$ has an infinite, but repeating pattern, of binary digits in the mantissa.
The computer must round or truncate this to fit within the finite number of bits available.
Therefore, the floating point representation of $0.1$ on a computer is merely an approximation.
The following cell illustrates that the floating point representation of $0.1$ is not exactly $0.1$.
Try entering some other numbers.

In [ ]:
print(0.1) # Python will (cleverly) print 0.1 as written
print(3*0.1) # We expect 0.3, but note there is a small error at the end!
print('{:.24f}'.format(0.1)) # This line asks Python to print 0.1 to 24 digits, don't worry about the details
print('{:.24f}'.format(0.1*3)) # The error in 3*0.1 is greater than 3 times the error of 0.1

The other type of number that you will use is an integer, and these generally work as you would expect.
For example, addition, multiplication, and integer powers of integers will result in another integer.
Once you start add a floating point number to a calculation the result will generally be converted to this type automatically.

You can force a conversion between integer and floating point types using `int()` and `float()`.
In the case of going from a floating point number to an integer the number will be truncated towards zero (which means rounded down if positive and rounded up if negative.

In [ ]:
print(int(4.2)) # rounded down
print(int(4.8)) # still rounded down
print(int(-4.2)) # rounded up
print(int(-4.8)) # still rounded up
print(float(23)) # simply adds '.0'
print(float(e)) # already a float

Another basic type is called a *boolean*, something which is either `True` or `False`.
For example, you can compare the two numbers using one of `<,>,<=,>=,==,!=` (here `!=` means not equal) and the result will be given as a boolean with value `True` or `False`.
Whilst not strictly a number type they can be used in calculation in which case `True` is treated as the number `1` and `False` is treated as the number `0`.

In [ ]:
print(3<10)
print(pi<e)
print(pi!=e)
print(3==3.0000)
print(2*True+False)

Note that generally one should avod comparing floating point numbers using `==` or `!=` as the result will often not be what you expect due to rounding errors.

In [ ]:
print(4.2==4.200000000000001)  # False as expected
print(4.2==4.2000000000000001) # True due to rounding of the latter

#### Exercise

 - Define a few variables to be equal to different numbers. 
 - Play around doing various calculations of these numbers via their variables.
 - Try a few different comparisons between different numbers.

In [ ]:
x = 2
y = x**2
z = 3*y - x


In [ ]:
math.sqrt(x**2 + y**2 + z**2)

In [ ]:
x < y < z

## Basic data containers in Python

Python has three basic data containers/structures.
A brief introduction to these is provided here.

### Lists

The first object we'll look at is a list.
A list is defined by using square brackets `[]` with a comma separated list of objects withn the brackets.
The objects within a list can be anything really, but mostly we will deal with lists of numbers.
Here are some examples.

In [ ]:
mylist = [1,2,e]
print(mylist)
anotherlist = [1,2,'skip a few',99,100] # this list includes a string!
print(anotherlist)
thirdlist = [mylist,anotherlist] # list containing two lists!
print(thirdlist)

The length of a list is the number of objects provided within those square brackets.
The length of a list can be determined via the `len()` function.
You can also use `len()` to get the length of many other objects.

In [ ]:
print(len(mylist))
print(len(anotherlist))

To access an object in a list you can pass its index.
You do this by adding square brackets `[]` immediately after the name of the list you want to get an object from, and within that you provide the index of the object you want.

**Important:** In Python, indexing starts with 0, that is the first object in a list has index 0 and the last object in a list has index one minus the length of the list.
A shortcut for accessing the last object is to use the index -1 (similarly -2 is the second last, and so on).
Here is an example.

In [ ]:
print(anotherlist[0]) # first object
print(anotherlist[2]) # the third object
print(anotherlist[len(anotherlist)-1]) # last object
print(anotherlist[-1]) # shortcut for the last object

If you ask for an object with index that does not exist you will get an error, for example:

In [ ]:
anotherlist[5] # raises IndexError

When you have a list within a list, you can access an obect of the inner most list by providing indices in succession, for example:

In [ ]:
print(thirdlist[0]) # gives the first object (mylist from above)
print(thirdlist[0][1]) # gives the second object in mylist

Objects in a list can be modified/changed at any time.
The list will *remember* this change.

In [ ]:
mylist[1] = -42
print(mylist) # the second object has changed
print(thirdlist) # this is also updated accordingly as it references mylist
thirdlist[1] = 0 # replace second object in thirdlist with number 0
print(thirdlist) 
print(anotherlist) # this remains unchanged though

You can add to a list by appending to it.
You can also *join* two lists by adding them together.

In [ ]:
mylist.append(4) # add the object '4' to end of mylist
print(mylist)
fourthlist = mylist + [5,6] # define a new list by joining mylist with [5,6]
print(fourthlist)

Long lists can be made using the `range()` function.
To be precise, `range()` is an *iterator* object, but can be converted to a list using `list()`.
If the variable `n` is an integer, then `range(n)` will iterate over the integers from 0 up to n-1. 
Therefore `list(range(n))` produces a list of the integers 0 up to n-1.
Observe this has length `n`. 

In [ ]:
longlist = list(range(100))
print(longlist)
print(len(longlist))
print(longlist[57]) # the 58'th object is 57 as we expect

Multiple objects in a list can be accessed using Python's slice notation.
Briefly, let `i,j,k` be indices with `i<j`, it works as follows:

 - `longlist[:i]` returns the first i objects (having indices 0 to i-1)
 - `longlist[i:]` returns the (i+1)th object and everything after it (having indices i to len(longlist)-1)
 - `longlist[i:j]` returns the (i+1)th object up to the j'th (having indices i up to j-1)
 - `longlist[::k]` returns every k'th object starting from the first (i.e. skipping every k-1 object in between)
 - 'longlist[i:j:k]' returns every k'th object starting from the i'th and not going beyong the (j-1)'th
 
Have a play around with the following examples to become familiar with how this works. 
Then, see if you can work out what `longlist[:j:k]` and `longlist[i::k]` do.
As a bonus, also try `longlist[::-1]` and `longlist[j:i:-k]`.

In [ ]:
print(longlist[:10]) # first 10 objects in list
print(longlist[95:]) # the 96th object and beyond
print(longlist[len(longlist)-10:]) # the last 10 objects
print(longlist[-10:]) # shortcut for last 10 objects
print(longlist[22:29]) # the 23rd to 29th objects (inclusive) 
print(longlist[::8]) # every eighth object stating from the first
print(longlist[34:49:5]) # every fifth object from the 35th up to the 49th
print(longlist[34:81:5]) # every fifth object from the 35th up to the 81st

Sometimes a copy of a list is desired.
One needs to take a little care with this when the original list is intended to remain unchanged.
I won't go through this in detail, there are some subtleties when your list contains more than just numbers, but you should take note of what happens in the following cells.

In [ ]:
integerlist = list(range(10))
copyoflist = integerlist # this is not really a 'copy'
copyoflist[4] = -1 # change a number
print(copyoflist)
print(integerlist) # Note this has also been changed

In [ ]:
integerlist = list(range(10))
copyoflist = integerlist[:] # This IS a copy
copyoflist[4] = -1
print(copyoflist)
print(integerlist) # Note this remains unchanged

In [ ]:
integerlist = list(range(10))
copyoflist = list(integerlist) # This IS a copy
copyoflist[4] = -1
print(copyoflist)
print(integerlist) # Note this remains unchanged

If you want to know more about lists in Python, I refer you to the documentation [here](https://docs.python.org/3/tutorial/datastructures.html).

Later you will learn about numpy arrays, which are a much more preferable data container/structure when dealing with lots of numbers.
However, lists can still be useful as an intermediate container when constructing arrays, particularly because lists are easily appended to (whereas this is more difficult, and less efficient, with arrays).
Indexing of numpy arrays also works the same way as lists so the tricks with slice notation are transferable.

#### Exercise

 - Define a list containing 5 numbers of your choosing. Print the 4th element in that list. Change the 2nd element to be equal to $42$ (without creating a new list). Add a 6th element to the list equal to $100$. Print the resulting list in reverse order.
 - Define a list containing the integers -10 up to 20 (including both ends). Print every 3rd element of that list starting with the 2nd element.
 

In [ ]:
l = [0,1,2,3,4]
print(l[3])
l[1]=42
l.append(100)
print(l[::-1])


In [ ]:
Z = list(range(-10,21,1))
print(Z[1::3])

### Tuples

A tuple is somewhat similar to a list.
It contains several elements which you enclose in round brackets/parentheses.
However, unlike a list, one a tuple is defined/constructed it can no longer be changed (in computer science terminology we would say a tuple is *immutable* whereas a list is *mutable*).
If a change to a tuple is required for some reason it essentially needs to be replaced with a new tuple.
There are select uses for tuples that you may see throughout the course, two examples being in describing the shape of multi-dimensional arrays, and occasionally when passing arguments to functions.

Tuples have a length which can be determined with `len()`, and are also indexed starting from zero the same as lists, and the same slice notation works.
Have a play with the examples below.

In [ ]:
mytuple = (1,2,3) # tuple containing 1,2,3
print(mytuple)
print(len(mytuple))
print(mytuple[1])
longertuple = tuple(range(10))
print(longertuple)
print(longertuple[3:8:2])
shorttuple = (5,) # tuple with one element, note the comma is important!
print(shorttuple)
notatuple = (5)
print(notatuple) # brackets here treated as if this was a calculation
print(type(shorttuple)) # use type() to query the type
print(type(notatuple))  # see, definitely not a tuple

Above you can see how the Python built-in function `type()` is used to determine exactly what type of object something is.

### Dictionaries

Dictionaries are a container/structure used to a construct lookup table which can be indexed by something other than integers. 
They have relatively little use in the context of this course but are still good to know about for select used.
They are constructed within curly brakcets `{}` and consist of *key:value* pairs, separated by the colon.
The *key* can be anything that is *hashable*, which includes integers and strings, and the *value* can be any python object.
A few examples are below.
You will learn a bit more if/when we use them.

In [ ]:
mydict = {'a':1,'b':2,'bob':3,'alice':33.7,-40:7.8} # construction
print(mydict)
print(mydict['bob']) # look up object associated with 'bob'
print(mydict[-40]) # look up object associated with 'bob'
mydict['sam'] # raises a KeyError as there is no such key in the dictionary

In [ ]:
mydict['sam'] = mylist # adds a new key,value pair, the value is a list in this case
print(mydict)

## Control flow in Python

There are several way to control the execution of code, generally referred to as 'control flow'.

### if/elif/else

The first is with if/else constructs. 
Here is how an if statement works.
```
if <statement>:
    <code that will execute is statement is True>
<code which will execute afterwards as per normal>
```
`<statement>` here can be anything that evaluates as `True` or `False`.
When the statement evaluates as `True` then the indented code that follows will be executed, but if the statement is evaluated as `False` then the indented code will be skipped.

The colon at the end of the `if` clause is important for making it clear the `<statement>` has ended.
Note that many lines of indented code can follow an `if` clause.
Code controlled by the `if` clause ends at the first line of code to appear which is not indented (or at the end of a cell).
There must be at least one indented line of executable code.

A common `<statement>` will be to compare two numbers, for example `a==b` will be `True` if `a` and `b` are equal, but will otherwise return `False` (recall the double equal sign is used to test for equality, as opposed to assigning a value to a variable with a single equals). 
Alternatively we can compare numbers using inequalities `a<b` or `a>b` or `a<=b` or `a>=b`.

In [ ]:
if 1+1==2:
    print('one plus one is two') # this will be executed
    
if 1+1==3:
    print('one plus one is three') # this will be skipped
    print('or is it?') # this will be skipped as well
    
if 1+1<4:
    print('one plus one is less than four')
    print('and is therefore less than 5 as well!') # also printed
print('hello') # executed as normal

Sometimes it is useful to some code executed when a statement is true, but different code to be executed when a statement is not true. We can use `else:` after an `if (statement):` clause for this.
`else:` appears on a single line and is has the same indentation level as the `if (statement):` it relates to (which for now means it is not indented).
Code that follows `else:`, which is to be executed when the `if` clause it relates to is `False`, must be indented. One can again include several lines of indented code.

In [ ]:
if 2*3==6:
    print('two times three is six') # will be executed
else:
    print('two times three is not six') # will not be executed
    
if 2*3==7:
    print('two times three is seven') # will not be executed
    print('or is it?') # also skipped
else:
    print('two times three is not seven') # will be executed
    print('it is six of course!') # also executed 
print('hello') # executed as normal

`elif (another statement):` can also be used after an initial `if (statement):`.
It will check `(another statement)` only if the original `(statement)` was `False`.
Indented code following `elif (another statement):` will only be executed if `(another statement)` is `True`.
Several `elif` clauses may follow an initial `if`, and an `else` clause may come after the final `elif`.

In [ ]:
if 1+1==0:
    print('one plus one is zero') # not exectuted, move onto next elif
elif 1+1==1:
    print('one plus one is one') # also not executed, move onto next elif
elif 1+1==2:
    print('one plus one is two') # is executed, any further elif/else will be ignored
elif 1+1==3:
    print('one plus one is three') # skipped
else:
    print('one plus one is none of the above') # skipped

You can "nest" if statements, consider the following example.
Change the value of k several times and examine the output.
Examine how code becomes further indented with the nested flow control statements.

In [ ]:
k=5
if k%2==0:
    print('k is a multiple of 2')
else:
    print('k is not a multiple of 2')
    if k%3==0:
        print('k is a multiple of 3')
    else:
        print('k is not a multiple of 3')

#### Exercises

 - Define $i$ to be some integer, then write down an if/else clause that checks if $i$ is a multiple of $7$ and prints an appropriate statement in either case. Test it on some different values of $i$.
 - Define $x$ to be some number, write an if/else clause that calculates $y=x^2$ if $x$ is positive but calculates $y=x^3$ otherwise. Test it on some different values of $x$.
 - Deifine $i$ to be some integer, then write an if/elif/else clause that calculates $j=i/3$ if $i$ is a multiple of $3$, calculates $j=2*i$ if $i$ has remainder $1$ upon division by $3$, and calculates $j=i+1$ if $i$ has remainder $2$ upon division by $3$.

In [ ]:
i = 3
if i%7 == 0:
    print(f"{i} is a multiple of 7")
else:
    print(f"{i} is not a multiple of 7")

In [ ]:
x = 3.14159
if x < 0:
    y=x**3
else:
    y=x**2
print(y)

In [ ]:
i = 3
if i%3==0:
    j=i/3
elif i%3 == 1:
    j = 2*i
elif i%3 == 2:
    j = i+1

print(f"{i}, {j}")

### Loops

There are two main mechanisms to loop code, `for` and `while`.

A `for` loop has the form
```
for <variable> in <list or iterator like object>:
    <code part of loop>
    <more code part of loop>
<normal code outside of loop>
```
Notice indentation again controls what is considered part of the loop.
Immediately after `for` we provide a `<variable>` which takes on different values on each iteration of the loop.
The values taken on by `<variable>` are described by the `<list or iterator like object>`.

Here are a several examples to consider.

In [ ]:
for i in [2,3,5,7]:
    print(i,' is prime')

In [ ]:
for a in range(10):
    print(a)    

In [ ]:
s=0
for i in range(5):
    s = s+i
print(s)

Note that `range(...)` can take 2 or 3 arguments to iterate over different sets of integers.
With two arguments $i<j$ then `range(i,j)` will produce the integers `i` up to `j-1` (note `j` itself is not included).
With three arguments `range(i,j,k)` will produce the sequence `i,i+k,i+2k,i+3k` and so on until it skips over `j` (note `j` will not be printed even if $i+mk$ is exactly $j$ for some integer $m\geq0$).
Negative $k$ can be provided along with $i>j$ to produce a sequence of numbers in reverse order.
Essentially, the three arguments work just like slice notation `[i:j:k]` when accessing a list.

Here are some examples to consider:

In [ ]:
for i in range(4,8):
    print(i)

In [ ]:
for i in range(3,10,2):
    print(i)

In [ ]:
for a in range(10,1,-3):
    print(a)

One can have multiple nested loops, or an if statement within a loop, or vice versa.

Study the following examples. One example below uses the operation `i%j` which calculates the remainder upon dividing $i$ by $j$ (with i,j being integers).

In [ ]:
for i in range(3):
    for j in range(2):
        print(i,j)

In [ ]:
for i in range(6):
    if i%2==0:
        print(i,' is even')
    else:
        print(i,' is odd')

A `while` loop has the form
```
while <statement>:
    <code as part of loop>
<code to execute as normal>
```
A while loop evaluates `<statement>`, if `True` it evaluates all the indented code and then returns to re-evaluate `<statement>`.
This repeats indefinitely until `<statement>` is evaluated to be `False`.

One must take care with `while` loops because if it is constructed in a way that the statement is never `False` then the loop will continue indefinitely (and it may appear that Python has "frozen").

Here are some examples to consider.

In [ ]:
k=0
while k<=5:
    k = k+1
print(k)

In [ ]:
kfactorial=1
k=5
while k>0:
    kfactorial = kfactorial*k
    k = k-1
print(kfactorial)

In [ ]:
a = True
k = 0
while a:
    if k>20:
        a = False
    k = k+1
print(k) # think about why this is 22 and not 21

A `break` statement may be used to exit a `for` or `while` loop early.
Here are some examples.

In [ ]:
for a in range(10):
    if a==7:
        break
    print(a) # note that the value 7 will not be printed

In [ ]:
i = 0
while True: 
    # Note: this loop will never end without a break statement
    if i>100:
        break
    i = i+1
print(i)

#### Exercises

 - Write down a loop which adds up the numbers $15$ up to $25$
 - Write down a loop which adds up the square of the numbers $1$ to $10$
 - Write down a loop which adds up the numbers $10$ to $40$ which are divisible by 7
 - Write down a loop which add the squares of the numbers $1$ to $20$ excluding any numbers which are divisible by $3$
 

In [ ]:
print(sum(range(15,25),0))

I = 0
for i in range(15,25):
    I += i
print(I)

In [ ]:
r = 0
for i in range(1,11):
    r += i**2
print(r)


In [ ]:
r = 0
for i in range (10,41):
    if i%7==0:
        r+=i
print(r)

In [ ]:
r = 0
for i in range (1,21):
    if i%3==0:
        r+=i**2
print(r)

#### Challenge question

The Collatz conjecture is that the following always terminates:

 - Start with any positive integer $n$
 - If $n$ is even then calculate $n/2$, otherwise calculate $3n+1$
 - Repeat the previous line on the new number indefinitely only ending the loop when the result is equal to $1$
 
Write a cell that first sets an initial number to $n$, then enters a while loop which carries out the operations described above ending when the result is equal to $1$. 
Try it out on several numbers $n$.

As an extension exercise, count the number of iterations performed until the loop terminates and print this once the loop finally exits.

In [ ]:
def collatz(start):
    count = 0
    n = start
    while n > 1:
        count+=1
        if n % 2:
            n = 3*n+1
        else:
            n /= 2
        
    return count

starts = list(range(0,1000))
counts = [collatz(start) for start in starts]
print(f"{len(starts)}, {len(counts)}")

import matplotlib.pyplot as plt
plt.plot(starts,counts)

### Other control flow

There are other forms of control flow, for example `try`/`catch` can be used to handle errors. 
We'll not consider any of these in this course though.

## Defining and using your own functions

Abstractly, a function takes one or more arguments as inputs, does some work and produces some output.
The arguments to a function in Python can be pretty much any object you can define in code, but in this course our function will generally take onr or more numbers or vectors as input, and produce a number or vector as output.

There are two ways to define a function. 
For short/simple functions we can use `lambda`.
Defining a function with `lambda` has the general form
```
<function name> = lambda <one or more variables>:<the result>
```
For example, the function $f(x)=x^2$ can be defined via
```
f = lambda x:x**2
```
A multi-variate function $g(x,y)=x-y^2$ can be defined via
```
g = lambda x,y:x-y**2
```
The following shows these examples and their subsequent usage.

In [ ]:
f = lambda x:x**2
g = lambda x,y:x-y**2
print(f(2))
print(g(3.14,-4.5))

More complex functions are defined via a `def <function name>(<arguments>):` construct.
Following this initial `def` including the function name and arguments you can have any number of indented lines which are executed whenever the function is called.
Generally you end a function definition with a `return <result>` statement to proved the result/output of the function back to the user where the function is called.

`def` is good to use when you have a complicated function that is more easily described over many lines of code.
A function defined using `def` is also free to used all of the control flow patterns you have learnt about previously.

Here are several examples (starting with the equivalent of the two lambda functions from earlier).

In [ ]:
def f(x):
    return x**2
def g(x,y):
    return x-y**2
print(f(2))
print(g(3.14,-4.5))

In [ ]:
def complicated_function(x,y,z):
    t = math.sin(x+y)
    s = math.cos(z-x)
    return 3*t-4*s**2
print(complicated_function(1,2,3))

In [ ]:
def signed_square(x):
    if x>=0:
        return x**2
    else:
        return -x**2
print(signed_square(2.0))
print(signed_square(-2.0))

There ar many subtleties that occur if you modify the arguments (ior nput variables) as part of the calculation within a function. The best practice is to avoid doing this (unless you know what you are doing).

Sometimes it is convenient for a function to have optional arguments.
This can be done with both `lambda` and `def` functions.

To specify an optional argument we simply need to assign it a default value to be used if the user does not provide one.
Optional arguments must come after all of the mandatory arguements when defining the function.
Here is an example of $g(x,y)$ from above where $y=0$ is used as default if it is not provided.

In [ ]:
g = lambda x,y=0:x-y**2
print(g(3.14))
print(g(3.14,-4.5))

def g(x,y=0):
    return x-y**2
print(g(3.14))
print(g(3.14,-4.5))

An existing function can be called within another function.
Consider the following example.

In [ ]:
def add_numbers(x,y):
    return x+y
def square_number(x):
    return x**2
def add_squares(x,y):
    x_squared = square_number(x)
    y_squared = square_number(y)
    return add_numbers(x_squared,y_squared)
print(add_squares(3,4))

#### Exercises

 - Define a function that takes $x$ as input and returns $y=x^2(1-x)$. Test it by calculating a few samples.
 - Define a function that takes $x$ and $y$ as input and returns $4-x**2-y**2+2*x*y$. Test it by calculating a few samples.
 - Define a function that calculates the factorial of a positive integer $m$. Test it by calculating a few samples.
 - Come up with some functions of your own to define and try out.

In [ ]:
def y(x):
    return (x**2)*(1-x)

import numpy as np
X = np.linspace(-10,10,100)
plt.plot(X, [y(x) for x in X])

In [ ]:
def f(x,y):
    return 4-(x**2)-(y**2)+(2*x*y)

In [ ]:
def factorial(n):
    if n < 0:
        return "No"
    fac = 1
    for i in range(1,n+1):
        fac = fac*i
    return fac

print(factorial(10))


## Matplotlib

Matplotlib is one of the main plotting libraries for Python.

We generally import it (in the first cell of a notebook) as follows:

In [ ]:
import matplotlib.pyplot as plt
%matplotlib inline

The line `%matplotlib inline` is a special line to use in jupyter notebook which will ensure plots display within the notebook rather than in a separate window (this should be the default behaviour on most up to date platforms, but specifying explicitly is a good idea).

Matplotlib can produce a large variety of plots in 2 and 3 dimensions, but in this lab we will concentrate on simple plots within the $xy$ plane.

`plt.plot(<list of x coordinates>,<list of y coordinates>)` will plot a sequence of lines between each $(x,y)$ coordinate pair corresponding to the lists (or later arrays) provided. 
We follow this with `plt.show()` to finalise and display the plot.

In [ ]:
xc = [1,2,3]
yc = [2,1,3]
plt.plot(xc,yc)
plt.show()

Matplotlib is clever and generally picks a display domain and range for the plot that fits all of the data you have provided.
You can however modify this by specifying the limits of the $x$ and $y$ axes manually.
We can also turn on a background grid.
For example:

In [ ]:
plt.plot(xc,yc)
plt.xlim(0.0,4.0)
plt.ylim(-1.0,5.0)
plt.grid()
plt.show()

The axis values shown, which also produce the grid, are also automatically chosen by matplotlib.
These too can be manually specified but for simplicity we'll just let matplotlib handle this most of the time.

Note that the axis values do not lie on the traditional x,y axes that pass through the origin, instead matplotlib places these on the bottom and left sides of the plot.
This is so the axes themselves don't get in the way of the data you are plotting.

We can add labels to each axis using `plt.xlabel(<string>)` and `plt.ylabel(<string>)`. 
A title can also be added to a plot via `plt.title(<string>)`.
These are demonstrated in the next cell.

Rather than plot lines between the points we can place markers at the points by specifying `'o'` as a third argument.
For example

In [ ]:
plt.plot(xc,yc,'o')
plt.xlim(0.0,4.0)
plt.ylim(-1.0,5.0)
plt.xlabel('the x axis')
plt.ylabel('the y axis')
plt.title('a plot of three points')
plt.grid()
plt.show()

If we want the markers in addition to dashed lines between the points then we can call the plot command twice using different arguments.
For example:

In [ ]:
plt.plot(xc,yc,'o')
plt.plot(xc,yc,'--')
plt.xlim(0.0,4.0)
plt.ylim(-1.0,5.0)
plt.grid()
plt.show()

Other possible markers are `'x'`,`'+'`,`'*'` (star),`'s'` (square) and many more.

Other possible line styles are `'-'` (normal line),`':'` (dotted),`'-.'` (dash-dotted), and many more

Observe that by having more than one `plt.plot` command before `plt.show()` they will appear on the same plot (or figure).
If we put a `plt.show()` after each individual `plt.plot` command they would appear on two separate plots (or figures).
Try it out!

Matplotlib automatically cycles through plot colors, but you can ask for specific colors by including it in the string which specifies the marker/line type.
Basic colors matplotlib knows about are `'r'` red, `'b'` blue, `'g'` green, `'c'` cyan, `'m'` magenta, `'y'` yellow, `''` black.
Arbitrary colors can be specified by other means that we will not discuss.

Here is an example

In [ ]:
plt.plot(xc,yc,'rs')
plt.plot(xc,yc,'g-.')
plt.xlim(0.0,4.0)
plt.ylim(-1.0,5.0)
plt.grid()
plt.show()

Okay, so how can we plot a function $y=f(x)$?

First we need to produce a sequence of $x$ coordinates, then calculate their corresponding $y$ coordinates, and then pass them to matplotlib.

Here is an example.

In [ ]:
x = list(range(-3,4)) # integers -3 to 3
f = lambda x:x**2     # function f(x)=x^2
y = [f(a) for a in x] # calculates f(a) for each a in the list of x values
plt.plot(x,y)
plt.grid()
plt.show()

You may notice this is not a very good plot.
The points are too far apart and we can see the straight lines between points.
To obtain a smooth looking plot we need many more x values closer together.

Here is one way you might approach this.

In [ ]:
i = list(range(-30,31)) # integers -30 to 30
x = [0.1*a for a in i]  # numbers -3,-2.9,-2.8,...,3.0
y = [f(a) for a in x]   # numbers f(a) for each a in the list x
plt.plot(x,y)
plt.grid()
plt.show()

This looks much nicer, but the above approach is not very efficient way to produce a smooth plot, both in terms of the amount of code required and the computation time.

It will be much easier and faster to use numpy arrays, which we are about to introduce.

But for the moment, try the following exercises.

#### Exercises

 - Produce a plot of the points $(-1,-1)$, $(-1,3)$, $(2,2)$, $(0,1)$, $(2,0)$. Plot both markers at the points and a dotted line through the points (in the order provided). Set the range of the plot to be $x\in[-4,4]$ and $y\in[-2,4]$. Add labels to the two axes.
 - Produce a plot of the function $f(x)=(x-3)x(x+3)$ for $x\in[-5,5]$. Ensure you choose enough sample points to get a smooth looking plot. You may let matplotlib choose the range automatically.

In [ ]:
x = [-1,-1,2,0,2]
y = [-1,3,2,1,0]
plt.plot(x,y,":")
plt.plot(x,y,".")
plt.xlim((-4,4))
plt.ylim((-2,4))

In [ ]:
import numpy as np

x = np.linspace(-5,5,1000)
y = (x-3)*x*(x+3)
plt.plot(x,y)

## Numpy

Numpy introduces arrays to Python.

There are several nice features to using arrays

 - They can be used to describe vectors and matrices
 - They provide an efficient to perform operations on many numbers
 - They make plotting smooth function simpler
 
There are many aspects to numpy and in this lab we will cover the basics of one dimensional arrays.

First though, the standard way to import numpy (usually in the first cell of a notebook) will be via:

In [23]:
import numpy as np
import matplotlib.pyplot as plt

### Array construction

There are several ways to construct one dimensional arrays.

 - `np.array(<list>)` converts a list of numbers into an array
 - `np.zeros(n)` constructs an array of length $n$ with each component equal to 0 ($n$ must be a positive integer)
 - `np.ones(n)` constructs an array of length $n$ with each component equal to 1.
 - `np.full(n,a)` constructs an array of length $n$ with each component equal to $a$ ($a$ can be any floating point number)
 - `np.arange(a,b,c)` constructs an array containing $a,a+c,a+2c,a+3c,\dots$ up to (but not including) $b$ (this behaves similar to `range(i,j,k)` but `a,b,c` can be any floating point number)
 - `np.linspace(a,b,n)` produces an array of length $n$ containing equally spaced samples from the interval $[a,b]$ (including the endpoints, so the spacing is $\frac{(b-a)}{(n-1)}$)
 
`np.linspace(a,b,n)` and `np.arange(a,b,c)` will generally be the most useful of these, especially for plotting.
Here is an some examples of each.

In [24]:
print(np.zeros(4))
print(np.ones(5))
print(np.full(3,-np.pi)) # np.pi is the number pi
print(np.arange(2,5,0.5))
print(np.linspace(0,10,6))

[0. 0. 0. 0.]
[1. 1. 1. 1. 1.]
[-3.14159265 -3.14159265 -3.14159265]
[2.  2.5 3.  3.5 4.  4.5]
[ 0.  2.  4.  6.  8. 10.]


The nice thing about numpy arrays is we can perform math operations on them as if they were a normal variable and that operation will be automatically applied to every component in the array.
The length of a one dimensional array can also be checked using `len(<array>)` or `<array>.size`.
Here are some examples.

In [25]:
v = np.linspace(0,5,6)
print(len(v))
print(v.size)
print(v)
print(2*v)
print(v**2)
print(v*(10-v)+2)

6
6
[0. 1. 2. 3. 4. 5.]
[ 0.  2.  4.  6.  8. 10.]
[ 0.  1.  4.  9. 16. 25.]
[ 2. 11. 18. 23. 26. 27.]


Not only is this convenient for coding, it is much faster then doing these oprations on each component in a loop.

Additionally, we can apply math functions like $\sin(x),\cos(x),\tan(x),e^x$ and so on to an array and it will be applied over each component.
Here are some examples.

In [26]:
v = np.linspace(0,np.pi,7) # the array [0,pi/6,pi/3,pi/2,2pi/3,5pi/6,1]
print(np.sin(v))
print(np.cos(v))
print(np.tan(v))
print(np.exp(v))

[0.00000000e+00 5.00000000e-01 8.66025404e-01 1.00000000e+00
 8.66025404e-01 5.00000000e-01 1.22464680e-16]
[ 1.00000000e+00  8.66025404e-01  5.00000000e-01  6.12323400e-17
 -5.00000000e-01 -8.66025404e-01 -1.00000000e+00]
[ 0.00000000e+00  5.77350269e-01  1.73205081e+00  1.63312394e+16
 -1.73205081e+00 -5.77350269e-01 -1.22464680e-16]
[ 1.          1.68809179  2.84965391  4.81047738  8.1205274  13.70819567
 23.14069263]


Observe here that mathematically $\tan(\pi/2)$ is not defined, but the nature of numerical approximation means that numpy returns $\tan(x)$ for $x$ very close to, but not exactly $\pi/2$.
The consequence is we get a very large number 1.63312394e+16 which is scientific notation for $1.63312394\times10^16$.

Similarly we expect $\sin(\pi)=0$, $\cos(\pi/2)=0$ and so on but we instead get a very small number like 1.22464680e-16 ($=1.22464680\times10^{-16}$).

Sometimes numpy will complain if you ask for something which is not sensible.
For example trying to take the natural log via `np.log` of zero.

In [27]:
np.log(v) # numpy complains because zero is in the array v

array([       -inf, -0.64702958,  0.0461176 ,  0.45158271,  0.73926478,
        0.96240833,  1.14472989])

We can calculate the sum or mean/average of all the components in an array via `np.sum()` and `np.mean()` respectively.

In [28]:
print(np.sum(v))
print(v.sum())  # a shortcut
print(np.mean(v))
print(v.mean()) # a shortcut

10.995574287564274
10.995574287564274
1.5707963267948963
1.5707963267948963


When you have different two arrays which are the same length then we can add them together, multiply them together, and so on. 
Each of these operations will be performed on a component by component basis.
If you try to do this with two arrays which are not the same length you will get an error.
You can also compute the dot product of two arrays of the same length using `np.dot()`.

In [29]:
u = np.linspace(0,1,6)
v = np.linspace(2,3,6)
print(u+v)
print(u*v)
print(v**u)
print(np.dot(u,v))

[2.  2.4 2.8 3.2 3.6 4. ]
[0.   0.44 0.96 1.56 2.24 3.  ]
[1.         1.17080491 1.41933364 1.77412415 2.27890607 3.        ]
8.2


#### Exercise

 - Produce an array containing the numbers $[0,1/4,1/2,3/4,\dots,2]$.
 - Calculate the cube of the array.
 - Produce a second array containing the numbers $[-4,-3,-2,\dots,4]$
 - Calculate the dot product of the second array with the cube of the first.

In [30]:
arr = np.arange(0,2+0.25,0.25)
cube = arr**3

In [31]:
arr2 = np.linspace(-4,4,9)
r = np.dot(arr2,cube)

### Plotting functions $y=f(x)$ using arrays

Okay, with this little bit of numpy and matplotlib knowledge it is now very easy to produce smooth plots of functions of the form $y=f(x)$.
We'll jump straight into some examples before proceeding with some exercises.

In [32]:
x = np.linspace(0,2*np.pi,10) # With only 10 samples this will be a rough looking plot
y = np.sin(x)
plt.plot(x,y)
plt.grid()
plt.show()

Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …

In [33]:
x = np.linspace(0,2*np.pi,100) # With 100 samples we get a smoother plot
y = np.sin(x)
plt.plot(x,y)
plt.grid()
plt.show()

In [34]:
x = np.arange(-5,5.1,0.1) # sample from -5 to 5 with spacing of 0.1 between points
y1 = (x-3)*x*(x+3)
y2 = 2*x
plt.plot(x,y1)
plt.plot(x,y2)
plt.grid()
plt.show()

#### Exercises

 - Plot $y=\tan(x)$ over $x\in[-3,3]$. You may want to restrict the $y$ range to get a better view of what happens around the origin.
 - Plot the function $y=x\sin(x^2)$ over $x\in[0,10]$
 - Plot the function $y=\frac{1}{10}x^3$ and $y=5-x$ on the same plot over $x\in[-5,5]$. Produce a second plot of the same functions but restrict the domain and range of the plot to "zoom in" over the area where the two curves intersect. Include a background grid in the plot and use this to estimate the coordinates of the intersection of the two functions.
 - Plot some other functions of your own choosing!

In [35]:
x = np.linspace(-5,5,1000)
plt.plot(x,np.tan(x),"o")
plt.ylim((-10,10))

(-10.0, 10.0)

In [36]:
x = np.linspace(0,10,1000)
y = x*np.sin(x**2)
plt.plot(x,y)

In [37]:
x = np.linspace(-5,5,1000)
y0 = 0.1*x**3
y1 = 5-x
plt.plot(x,y0,x,y1)
plt.grid()
plt.show()

plt.plot(x,y0,x,y1)
plt.ylim((2,2.5))
plt.xlim((2.7,2.9))
plt.grid()
plt.show()
print("(2.8,2.2)")


(2.8,2.2)


## Plotting parametrised curves

The way that `plt.plot` works makes it very easy to plot parametrised curves of the form $(x,y)=(f(t),g(t))$.

We just need to start by constructing an array of samples of $t$ on some interval, then calculate the arrays $x=f(t)$ and $y=g(t)$, and lastly pass these to the `plt.plot()` function as per normal.

For example, we can plot some very nice Lissajous curves.

In [38]:
t = np.linspace(0,2*np.pi,200)
x = np.cos(t)
y = np.sin(3*t)
plt.plot(x,y)
plt.show()

#### Exercises

We'll play around with parametrised curves a lot more in a future lab, but for now, have a go at the following.

 - Plot the cardioid curve described in the course notes
 - Plot some other parametrised curves, come up with some whacky $f(t),g(t)$ functions of your own!

In [39]:
t = np.linspace(0,2*np.pi,1000)
x = (1-np.cos(t))*np.cos(t)
y = (1-np.cos(t))*np.sin(t)
plt.plot(x,y)

In [ ]:
%matplotlib widget
t = np.linspace(0,2*np.pi,10000)
plt.plot(np.cos(9*t),np.sin(10*t))